File: 2018/day02.go
package main

import (
	"fmt"
	"os"
	"strings"
	"io/ioutil"
)

func find(count int, m map[rune]int) bool {
	for _, v := range m {
		if v == count {
			return true
		}
	}
	return false
}


func part1(input string) interface{} {
    parts := strings.Split(input, "\n")
    twoCount := 0
    threeCount := 0
    for _, part := range parts {
    	letterCount := make(map[rune]int)
    	runes := []rune(part)
		for _, r := range runes {
			v, ok := letterCount[r]
			if ok {
				letterCount[r] = v + 1
			} else {
				letterCount[r] = 1
			}
		}
		if find(2, letterCount) {
			twoCount++
		}
		if find(3, letterCount) {
			threeCount++
		}
    }
    return twoCount * threeCount
}

func part2(input string) interface{} {
	parts := strings.Split(input, "\n")
	letterCountMap := make(map[string]map[rune]int)
	for _, part := range parts {
		letterCount := make(map[rune]int)
		runes := []rune(part)
		for _, r := range runes {
			v, ok := letterCount[r]
			if ok {
				letterCount[r] = v + 1
			} else {
				letterCount[r] = 1
			}
		}
		letterCountMap[part] = letterCount
	}
	return findAnswer(0, parts, letterCountMap)
}

func findAnswer(index int, parts []string, letterCountMap map[string]map[rune]int) string {
	if index == len(parts) - 1 {
		fmt.Println("Could not find an answer")
		os.Exit(1)
	}
	part1 := parts[index]
	letterCount1 := letterCountMap[part1]
	for i := index + 1; i < len(parts); i++ {
		part2 := parts[i]
		letterCount2 := letterCountMap[part2]
		failed := 0
		for k, v1 := range letterCount1 {
			v2, ok := letterCount2[k]
			if ok && v1 != v2 {
				failed++
			}
			if failed > 1 {
				break
			}
		}
		if failed < 2 {
			runes1 := []rune(part1)
			runes2 := []rune(part2)
			runes3 := []rune{}
			for i := 0; i < len(runes1); i++ {
				if (runes1[i] == runes2[i]) {
					runes3 = append(runes3, runes1[i])
				}
			}
			if len(runes3) == len(runes1) - 1 {
				return string(runes3)
			}
		}
	}
	return findAnswer(index + 1, parts, letterCountMap)
}

func assert(f func(string)interface{}, input string, expected interface{}) {
	result := f(input)
	if result != expected {
		fmt.Printf("Assert failed: got [%d] expecting [%d]\n", result, expected)
		os.Exit(1)
	}	
}

func readFile(path string) string {
    b, err := ioutil.ReadFile(path)
    if err != nil {
    	fmt.Printf("Error reading file: %s\n", err)
    	os.Exit(1)
    }
    return string(b)
}

func main() {
	i1 := `abcdef
bababc
abbcde
abcccd
aabcdd
abcdee
ababab`

	assert(part1, i1, 12)

	data := readFile("day02.in")
	fmt.Println(part1(data))

	i2 := `abcde
fghij
klmno
pqrst
fguij
axcye
wvxyz`

	assert(part2, i2, "fgij")
	fmt.Println(part2(data))
}

File: 2019/day02.kt
import java.io.File

fun part1(input: String): List<Int> {
    var c = input.split(",").map{ it.toInt() } as MutableList<Int>
    return _part1(c)
}

fun _part1(c: MutableList<Int>): List<Int> {
    var p = 0
    while (c[p].toInt() != 99) {
        var jump = 0
        when (c[p]) {
            1 -> {
                c[c[p + 3]] = c[c[p + 1]] + c[c[p + 2]]
                jump = 4
            }
            2 -> { 
                c[c[p + 3]] = c[c[p + 1]] * c[c[p + 2]]
                jump = 4
            }
            else -> {
            }
        }
        p = p + jump
    }
    return c
}

fun main() {
    assert(part1("1,9,10,3,2,3,11,0,99,30,40,50")[0] == 3500)
    assert(part1("1,0,0,0,99")[0] == 2)
    assert(part1("2,3,0,3,99")[3] == 6)
    assert(part1("2,4,4,5,99,0")[5] == 9801)
    assert(part1("1,1,1,4,99,5,6,0,99")[0] == 30)

    var lines = File("day02.in").readLines()
    var c = lines[0].split(",").map{ it.toInt() } as MutableList<Int>
    c[1] = 12
    c[2] = 2
    var result = _part1(c)
    println(result[0])

    for (noun in 0..99) {
        for (verb in 0..99) {
            c = lines[0].split(",").map{ it.toInt() } as MutableList<Int>
            c[1] = noun
            c[2] = verb
            if (_part1(c)[0] == 19690720) {
                println(100 * noun + verb)
                return
            }
        }
    }
}
